### ISSUE TITLE
parseJWTWithClaim allocation heavy

### ISSUE THREAD

**Issue by klauspost** [minio/minio#8799]
On benchmarks with 2 servers and uploading/getting 1MB blocks `parseJWTWithClaims` is responsible for more than 20% of all allocations. Most of it is related to JSON decoding.

Drilldown:

![mem](https://user-images.githubusercontent.com/5663952/72224139-16e97380-3577-11ea-8c72-f0f4ba4476fc.png)

In terms of CPU, it is using ~14% (of total server CPU) of which 11% (of total server CPU) is because of allocation.

**Describe the solution you'd like**

Ideally verifying tokens should not generate this much garbage. Since this functionality affects all calls and RPC this could be a significant improvement if improved.

**Describe alternatives you've considered**

I don't much like rolling our own, but this is quite dirty for what it is doing.

**Comment by harshavardhana** [comment]
Let me verify if this a quick fix here, did you verify this using warp with profiling? @klauspost

**Comment by klauspost** [comment]
@harshavardhana Yes. I looked if there was anything in the jwt library. There is a v4 coming, but it doesn't seem to be performance oriented.

It seems to me we would be looking at changing to another package or doing something ourselves. We don't use any fancy features of JWT, so we could do a simpler format.

If you find something that would be great. Otherwise I can take a look later.

**Comment by harshavardhana** [comment]
I found some improvements @klauspost let me push a PR and see if it improves @klauspost

**Comment by klauspost** [comment]
I don't think we can get rid of allocations unless we either write a new JWT parser or switches to something that isn't based on JSON.

I don't suppose there is anything preventing us from changing the tokens to something other than JWT that supports the same features?

**Comment by harshavardhana** [comment]
No we cannot do that @klauspost jwt is what we want for claims and hmac we didn't wish to have many ways to do the same thing since we need to support STS tokes based on jwt as well.

**Comment by harshavardhana** [comment]
I can see if we can change it such that we use typed claims v/s map claims.

**Comment by aead** [comment]
The performance issue here is caused by the JWT parsing in `storageServerRequestValidate`. However, the actual root cause here is not JWT but the usage of JWT as authentication mechanism between the storage nodes.

Instead of trying to (further) optimize the JWT implementation (with all its potential downsides - e.g. maintaining a fork a.s.o) I propose to use a more suited authentication mechanism for inter-node communication.

In particular, I suggest to use mTLS. So each minio node will generate 2 TLS private key / certificate pairs on startup and uses the private/public keys to establish a secure connection via TLS when connecting to another MinIO node.

I suggest to derive the TLS private keys from the admin access key and secret key. In particular:
```
seed_server = HMAC(secret_key, access_key || "server")
seed_client   = HMAC(secret_key, access_key || "client")

privateKey_server = ed25519.NewKeyFromSeed(seed_server)
privateKey_client = ed25519.NewKeyFromSeed(seed_client)
```
Observe that each server node computes the same `privateKey_server` and `privateKey_client` **iff** they share the same admin access key / secret key pair. 

Now, we can configure the TLS stack of the storageRestServer to present `publicKey_server` and require `publicKey_client` from its connection peer - and storageRestClient  to present `publicKey_client` and require `publicKey_server` from its connection peer.

This way we deterministically derive the TLS private/public keys on each server node from the admin access key / secret key and use TLS for authentication + inter-node traffic privacy. /cc @harshavardhana @klauspost

**Comment by harshavardhana** [comment]
> In particular, I suggest to use mTLS. So each minio node will generate 2 TLS private key / certificate pairs on startup and uses the private/public keys to establish a secure connection via TLS when connecting to another MinIO node.

I like the idea @aead - let me give this a quick spin.

**Comment by harshavardhana** [comment]
The challenge with mTLS is @aead that we need to open a separate port for API calls and internal calls.  This would cause a lot of hindrance to make it work with our existing style. 

If there is a possibility of just using this for internal API operations using net/http we should do it. But I am not able to see how it is possible. This is why we needed a bearer token style (JWT) authentication schemes.

**Comment by aead** [comment]
Hm, yeah I missed the limitation of a single port. In case of plain HTTP we would need to handle S3 API requests over plain HTTP while inter-node communication over TLS uses the same port - which does not work / would be really messy.  
That's unfortunate. 

Okay, then let me propose a much simpler fix.
At the moment we have a expiry for inter-node JWT tokens of [~100 years](https://github.com/minio/minio/blob/3011692d9347880e14002cf36fc4809cc77c5552/cmd/jwt.go#L39).
Since a inter-node JWT token won't expire in practice I suggest to:
1. Parse the token and verify it when we receive it the first time.
2. If the token is valid "cache" it [here](https://github.com/minio/minio/blob/3011692d9347880e14002cf36fc4809cc77c5552/cmd/storage-rest-server.go#L86) by adding a field `prevToken string` to [`storageRestServer`](https://github.com/minio/minio/blob/3011692d9347880e14002cf36fc4809cc77c5552/cmd/storage-rest-server.go#L42)
3. Whenever we receive a inter-node request we first compare the token in the request to the cached one (`prevToken`) using `subtle.ConstantTimeCompare` and only parse and verify the token again if the received one does not match the cached one.

This way we will reduce the number of JWT parsing and verification dramatically with a minimal change to the existing code base. For extracting the token we can use:
`jwt.AuthorizationHeaderExtractor.ExtractToken(req)`

AFAICS this should not weaken the current security but reduce the number of JWT token verifications.
@harshavardhana

**Comment by harshavardhana** [comment]
https://github.com/minio/minio/pull/8802 the PR already does this, but yes with a good amount of changes especially it adds more security for the token themselves per API call like S3 API 15minute validity.

**Comment by harshavardhana** [comment]
The problem has been security authenticity of these changes so your input on the PR would be helpful

**Comment by aead** [comment]
I'm not able to see that we cache the token. My point being that if we do that all the other changes won't be necessary since JWT token parsing / verification won't be a hot path anymore?!

In particular: https://github.com/minio/minio/pull/8802/files#diff-9c5604c6beb0f12c48e5615563d70c97R59 Here we should remember the token string...

**Comment by aead** [comment]
```
diff --git a/cmd/storage-rest-server.go b/cmd/storage-rest-server.go
index 325bc32e..84634298 100644
--- a/cmd/storage-rest-server.go
+++ b/cmd/storage-rest-server.go
@@ -18,6 +18,7 @@ package cmd
 
 import (
    "context"
+   "crypto/subtle"
    "encoding/gob"
    "encoding/hex"
    "errors"
@@ -30,6 +31,7 @@ import (
    "strings"
    "time"
 
+   jwtreq "github.com/dgrijalva/jwt-go/request"
    "github.com/gorilla/mux"
    "github.com/minio/minio/cmd/config"
    xhttp "github.com/minio/minio/cmd/http"
@@ -40,7 +42,8 @@ var errDiskStale = errors.New("disk stale")
 
 // To abstract a disk over network.
 type storageRESTServer struct {
-   storage *posix
+   storage       *posix
+   prevAuthToken string
 }
 
 func (s *storageRESTServer) writeErrorResponse(w http.ResponseWriter, err error) {
@@ -80,10 +83,23 @@ func storageServerRequestValidate(r *http.Request) error {
 
 // IsValid - To authenticate and verify the time difference.
 func (s *storageRESTServer) IsValid(w http.ResponseWriter, r *http.Request) bool {
-   if err := storageServerRequestValidate(r); err != nil {
-       s.writeErrorResponse(w, err)
+   token, err := jwtreq.AuthorizationHeaderExtractor.ExtractToken(r)
+   if err != nil {
+       if err == jwtreq.ErrNoTokenInRequest {
+           s.writeErrorResponse(w, errNoAuthToken)
+           return false
+       }
+       s.writeErrorResponse(w, errNoAuthToken)
        return false
    }
+   if s.prevAuthToken == "" || subtle.ConstantTimeCompare([]byte(token), []byte(s.prevAuthToken)) != 1 {
+       if err := storageServerRequestValidate(r); err != nil {
+           s.writeErrorResponse(w, err)
+           return false
+       }
+       s.prevAuthToken = token
+   }
+
    diskID := r.URL.Query().Get(storageRESTDiskID)
```

**Comment by harshavardhana** [comment]
The point was actually not to cache but instead provide token per API call such that the query params can be validated properly and securely.

Current approach is actually a remnant of Go's net/rpc code which we used to have where we had a single token for the single transport, it doesn't make sense for HTTP API.

So the PR itself is making such changes to ensure that jwt also validates the query params which is basically the payload for the http internode APIs.

So rather than caching it is perhaps better to implement this in a more secure way. PTAL at the PR itself.

Currently I am of the opinion that we should take it in, if it doesn't have any security issues.

@klauspost also I would suggest if you can provide feedback on what can be improved interms of readability about mapClaims v/s standardClaims

If we can get new numbers with the change would be beneficial to see what all areas might see improvement. If it's 50% improvement it is still a lot when the workload is lots of concurrent small file operations.

**Comment by aead** [comment]
> provide token per API call such that the query params can be validated properly and securely.

That is a nice improvement :+1: 
 
> So rather than caching it is perhaps better to implement this in a more secure way. PTAL at the PR itself.

The PR itself looks good to me. It's just that benchmarks have to proof that it actually resolves the performance issues described above. If it turns out that it reduces the JWT verification cost sufficiently, then I'm more than okay with dropping the caching idea.

**Comment by harshavardhana** [comment]
Can this issue be closed @klauspost now that #8802 is merged?

**Comment by harshavardhana** [comment]
Closing this as fixed and released #8802
